src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Endpoints/Completion.cs

```diff
@@ -271,4 +282,135 @@ private static TextEdit BuildRevertedEdit(TextEdit provisionalTextEdit)
         var formattingOptions = _formattingOptionsProvider.GetOptions(document.TextDocumentIdentifier.Uri);
         return Task.FromResult(formattingOptions);
     }
+
+    [JsonRpcMethod(LanguageServerConstants.RazorSnippetCompletionEndpointName, UseSingleObjectParameterDeserialization = true)]
+    public CompletionItem[]? GetSnippetCompletions(RazorSnippetCompletionParams completionParams, CancellationToken _)
+    {
+        if (completionParams.TextSpan.Length == 0)
+        {
+            return null;
+        }
+
+        var snippets = _snippetCache.GetSnippets();
+
+        if (snippets.IsDefaultOrEmpty)
+        {
+            return null;
+        }
+
+        var langSpecificCompletionsEnum = completionParams.LanguageKind switch
+        {
+            RazorLanguageKind.Html => snippets.Where(static s => s.Language == SnippetLanguage.Html),
+            RazorLanguageKind.CSharp => snippets.Where(static s => s.Language == SnippetLanguage.CSharp),
+            _ => null
+        };
+
+        if (langSpecificCompletionsEnum is null)
+        {
+            return null;
+        }
+
+        if (!_documentManager.TryGetDocument(completionParams.Identifier.Uri, out var snapshot))
+        {
+            return null;
+        }
+
+        var text = snapshot.Snapshot.GetText(completionParams.TextSpan.Start, completionParams.TextSpan.Length);
+
+        var completions = langSpecificCompletionsEnum
+            .Where(s => s.Shortcut.StartsWith(text, StringComparison.OrdinalIgnoreCase))
```

Comment:
Do we normally do this sort of filtering? I thought the LSP client did the filtering, and we just return all of the items we know about. I _think_ once a completion session is started, we don't get called for each keystroke, so if `text` is "ta" when the session starts, we'll add the "table" snippet, but not "th", and if the user backspaces to just have "t", they won't see the "th" snippet.

Reply:
Reading the spec, we would do this if we return incomplete results. We don't normally, so I updated to return all snippets for a given language

Reply:
> Reading the spec

You now officially know the most about completion :)